spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -293,28 +307,130 @@ private String evaluateTopic(Object request, Object source, Object result, Expre
+	/**
+	 * Send the result to the topic.
+	 *
+	 * @deprecated in favor of {@link #sendResponse(Object, String, Object)}.
+	 * @param result the result.
+	 * @param topic the topic.
+	 */
+	@Deprecated
 	protected void sendResponse(Object result, String topic) {
+		sendResponse(result, topic, null);
+	}
+
+	/**
+	 * Send the result to the topic.
+	 *
+	 * @param result the result.
+	 * @param topic the topic.
+	 * @param source the source (input).
+	 * @since 2.1.3
+	 */
+	@SuppressWarnings("unchecked")
+	protected void sendResponse(Object result, String topic, @Nullable Object source) {
 		if (topic == null) {
 			if (this.logger.isDebugEnabled()) {
 				this.logger.debug("No replyTopic to handle the reply: " + result);
 			}
 		}
+		else if (result instanceof Message) {
+			this.replyTemplate.send((Message<?>) result);
+		}
 		else {
 			if (result instanceof Collection) {
 				((Collection<V>) result).forEach(v -> {
 					this.replyTemplate.send(topic, v);
 				});
 			}
 			else {
-				this.replyTemplate.send(topic, (V) result);
+				byte[] correlationId = null;
+				boolean sourceIsMessage = source instanceof Message;
+				if (sourceIsMessage
+						&& ((Message<?>) source).getHeaders().get(KafkaHeaders.CORRELATION_ID) != null) {
+					correlationId = ((Message<?>) source).getHeaders().get(KafkaHeaders.CORRELATION_ID, byte[].class);
+				}
+				if (correlationId == null) {
+					correlationId = tryCorrelationIdFallback(source);
+				}
+				if (sourceIsMessage) {
+					MessageBuilder<Object> builder = MessageBuilder.withPayload(result)
+							.setHeader(KafkaHeaders.TOPIC, topic);
+					if (correlationId != null) {
+						builder.setHeader(KafkaHeaders.CORRELATION_ID, correlationId);
+					}
+					setPartition(builder, ((Message<?>) source));
+					this.replyTemplate.send(builder.build());
+				}
+				else {
+					this.replyTemplate.send(topic, (V) result);
+				}
+			}
+		}
+	}
+
+	private void setPartition(MessageBuilder<Object> builder, Message<?> source) {
+		byte[] partitionBytes = source.getHeaders().get(KafkaHeaders.REPLY_PARTITION, byte[].class);
+		if (partitionBytes == null) {
+			partitionBytes = tryPartitionFallback(source);
+		}
+		if (partitionBytes != null) {
+			String partition = new String(partitionBytes, StandardCharsets.UTF_8);
+			builder.setHeader(KafkaHeaders.PARTITION_ID, Integer.parseInt(partition));
+		}
+	}
+
+	@Nullable
+	private byte[] tryTopicFallback(Object source) {
+		return extractNativeHeader(source, KafkaHeaders.REPLY_TOPIC);
+	}
+
+	@Nullable
+	private byte[] tryPartitionFallback(Object source) {
+		return extractNativeHeader(source, KafkaHeaders.REPLY_PARTITION);
+	}
+
+	@Nullable
+	private byte[] tryCorrelationIdFallback(Object source) {
+		return extractNativeHeader(source, KafkaHeaders.CORRELATION_ID);
+	}
+
+	/*
+	 * Needed if Jackson is not on the classpath since we can't map headers in that case.
+	 */
+	private byte[] extractNativeHeader(Object source, String header) {
```

Comment:
Don't merge yet; I have found a better solution for when Jackson is not on the CP.